File: packages/php/src/Types/Metadata/JatsMetadata.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types\Metadata;

/**
 * JATS (Journal Article Tag Suite) metadata.
 *
 * Contains copyright, license, publication history dates,
 * and contributor role information.
 */
readonly class JatsMetadata
{
    /**
     * @param array<string, string> $historyDates
     * @param ContributorRole[] $contributorRoles
     */
    public function __construct(
        public ?string $copyright = null,
        public ?string $license = null,
        public array $historyDates = [],
        public array $contributorRoles = [],
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $rawCopyright = $data['copyright'] ?? null;
        $copyright = is_string($rawCopyright) ? $rawCopyright : null;

        $rawLicense = $data['license'] ?? null;
        $license = is_string($rawLicense) ? $rawLicense : null;

        /** @var array<string, string> $historyDates */
        $historyDates = $data['history_dates'] ?? [];
        if (!is_array($historyDates)) {
            $historyDates = [];
        }

        /** @var ContributorRole[] $contributorRoles */
        $contributorRoles = [];
        $rawRoles = $data['contributor_roles'] ?? null;
        if (is_array($rawRoles)) {
            foreach ($rawRoles as $roleData) {
                if (is_array($roleData)) {
                    /** @var array<string, mixed> $typedRoleData */
                    $typedRoleData = $roleData;
                    $contributorRoles[] = ContributorRole::fromArray($typedRoleData);
                }
            }
        }

        return new self(
            copyright: $copyright,
            license: $license,
            historyDates: $historyDates,
            contributorRoles: $contributorRoles,
        );
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $result = [];

        if ($this->copyright !== null) {
            $result['copyright'] = $this->copyright;
        }

        if ($this->license !== null) {
            $result['license'] = $this->license;
        }

        if ($this->historyDates !== []) {
            $result['history_dates'] = $this->historyDates;
        }

        if ($this->contributorRoles !== []) {
            $result['contributor_roles'] = array_map(
                fn (ContributorRole $c) => $c->toArray(),
                $this->contributorRoles,
            );
        }

        return $result;
    }
}

File: packages/elixir/lib/kreuzberg/jats_metadata.ex
defmodule Kreuzberg.JatsMetadata do
  @moduledoc """
  JATS (Journal Article Tag Suite) metadata.

  Matches the Rust `JatsMetadata` struct.

  ## Fields

    * `:copyright` - Copyright statement
    * `:license` - License information
    * `:history_dates` - Publication history dates (e.g., %{"received" => "2024-01-01"})
    * `:contributor_roles` - List of contributors with roles
  """

  @type t :: %__MODULE__{
          copyright: String.t() | nil,
          license: String.t() | nil,
          history_dates: map(),
          contributor_roles: list(Kreuzberg.ContributorRole.t())
        }

  defstruct copyright: nil,
            license: nil,
            history_dates: %{},
            contributor_roles: []

  @spec from_map(map()) :: t()
  def from_map(data) when is_map(data) do
    contributor_roles =
      (data["contributor_roles"] || [])
      |> Enum.map(fn
        %Kreuzberg.ContributorRole{} = cr -> cr
        map when is_map(map) -> Kreuzberg.ContributorRole.from_map(map)
        _ -> nil
      end)
      |> Enum.reject(&is_nil/1)

    %__MODULE__{
      copyright: data["copyright"],
      license: data["license"],
      history_dates: data["history_dates"] || %{},
      contributor_roles: contributor_roles
    }
  end

  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = meta) do
    %{
      "copyright" => meta.copyright,
      "license" => meta.license,
      "history_dates" => meta.history_dates,
      "contributor_roles" => Enum.map(meta.contributor_roles, &Kreuzberg.ContributorRole.to_map/1)
    }
  end
end
